src/mesh/p4est_mesh.jl

```diff
@@ -224,3 +350,50 @@ function calc_node_coordinates!(node_coordinates, mapping, trees_per_dimension,
     end
   end
 end
+
+
+# Calculate physical coordinates of each node.
+# Extract corners of each tree, interpolate to requested interpolation nodes,
+# map the resulting coordinates with the specified mapping.
+# 2D version
+function calc_tree_node_coordinates!(node_coordinates::AbstractArray{RealT, 4},
+                                     nodes, mapping,
+                                     vertices, tree_to_vertex) where RealT
+  nodes_in = [-1.0, 1.0]
+  matrix = polynomial_interpolation_matrix(nodes_in, nodes)
+  data_in = Array{RealT, 3}(undef, 2, 2, 2)
+  tmp1 = zeros(RealT, 2, length(nodes), length(nodes_in))
+
+  for tree in 1:size(tree_to_vertex, 2)
+    # Tree vertices are stored in Z-order, ignore z-coordinate in 2D, zero-based indexing
+    data_in[:, 1, 1] .= vertices[1:2, tree_to_vertex[1, tree] + 1]
+    data_in[:, 2, 1] .= vertices[1:2, tree_to_vertex[2, tree] + 1]
+    data_in[:, 1, 2] .= vertices[1:2, tree_to_vertex[3, tree] + 1]
+    data_in[:, 2, 2] .= vertices[1:2, tree_to_vertex[4, tree] + 1]
+
+    multiply_dimensionwise!(
+      view(node_coordinates, :, :, :, tree),
+      matrix, matrix,
+      data_in,
+      tmp1
+    )
+  end
+
+  map_node_coordinates!(node_coordinates, mapping)
+end
+
+function map_node_coordinates!(node_coordinates::AbstractArray{RealT, 4}, mapping) where RealT
+  for tree in axes(node_coordinates, 4),
+      j in axes(node_coordinates, 3),
+      i in axes(node_coordinates, 2)
+
+    node_coordinates[:, i, j, tree] .= mapping(node_coordinates[1, i, j, tree],
+                                                 node_coordinates[2, i, j, tree])
```

Comment:
```suggestion
    node_coordinates[:, i, j, tree] .= mapping(node_coordinates[1, i, j, tree],
                                               node_coordinates[2, i, j, tree])
```
I wonder whether we should switch to `mapping`s that accept the coordinates as a single `SVector`/`Tuple`/`AbstractVector`/iterable... Then, we could just use `identity` for the identity mapping... However, this won't block this PR and we might continue this discussion elsewhere.

Reply:
If we change this, we should change this everywhere in the `CurvedMesh` as well.

Reply:
Yes, that was more a general kind of comment. Please add it to an issue to discuss it there and resolve this thread.

Reply:
Discussion moved to #602